python/ql/test/experimental/import-resolution/importflow.ql

```diff
@@ -0,0 +1,122 @@
+import python
+import semmle.python.dataflow.new.DataFlow
+import semmle.python.ApiGraphs
+import TestUtilities.InlineExpectationsTest
+import semmle.python.dataflow.new.internal.ImportResolution
+
+/** A string that appears on the right hand side of an assignment. */
+private class SourceString extends DataFlow::Node {
+  string contents;
+
+  SourceString() {
+    this.asExpr().(StrConst).getText() = contents and
+    this.asExpr().getParent() instanceof Assign
+  }
+
+  string getContents() { result = contents }
+}
+
+/** An argument that is checked using the `check` function. */
+private class CheckArgument extends DataFlow::Node {
+  CheckArgument() { this = API::moduleImport("trace").getMember("check").getACall().getArg(1) }
+}
+
+/** A data-flow node that is a reference to a module. */
+private class ModuleRef extends DataFlow::Node {
+  Module mod;
+
+  ModuleRef() {
+    this = ImportResolution::getModuleReference(mod) and
+    not mod.getName() in ["__future__", "trace"]
+  }
+
+  string getName() { result = mod.getName() }
+}
+
+/**
+ * A data-flow node that is guarded by a version check. Only supports checks of the form `if
+ *sys.version_info[0] == ...` where the right hand side is either `2` or `3`.
+ */
```

Comment:
This is very specific. But its use seems even more specific, namely only those where `version` is `3`. Are you keeping the option open for a `ResolutionTest2`? Otherwise this node should perhaps be `Python3Node`...unless the point is to exclude code guarded by version 2 from all tests?

Reply:
Yeah, I figured there was no harm in leaving in the possibility of extending it to also handle Python 2-only tests. However, as I was not aware of any Python 2 behaviour that needed testing, I decided not to implement it now. (In part also because I had only just gotten this to work reliably, after trying a whole bunch of variations that failed for one reason or another.)

If we end up needing Python 2 stuff, I think the better approach would be to simply rename `ResolutionTest3` to `VersionSpecificResolutionTest` and use `major_version()` to decide which tag is active and which tag to look for in the actual results. That way, we don't have to (basically) duplicate the body of `hasActualResult`.

Reply:
Yeah, I figured there was no harm in leaving in the possibility of extending it to also handle Python 2-only tests. However, as I was not aware of any Python 2 behaviour that needed testing, I decided not to implement it now. (In part also because I had only just gotten this to work reliably, after trying a whole bunch of variations that failed for one reason or another.)

If we end up needing Python 2 stuff, I think the better approach would be to simply rename `ResolutionTest3` to `VersionSpecificResolutionTest` and use `major_version()` to decide which tag is active and which tag to look for in the actual results. That way, we don't have to (basically) duplicate the body of `hasActualResult`.